File: src/Extract.jl
using ArgParse
using Distributed

@everywhere using DataFrames
@everywhere using Statistics
@everywhere using Printf
@everywhere using GZip
@everywhere using CSV

@everywhere const re_MolName = r"^#+ +Name: +"
@everywhere const re_Number = r"^#+ +Number: +"
@everywhere const re_OXR = r"^#+ +OXR +"
@everywhere const re_Total = r"^#+ +Total Energy: +"

@everywhere function build_gzfn(sub_idx, cls_idx)
    # zero padded cls_idx
    zp_idx = @sprintf "%04d" parse(Int64, cls_idx)

    mol2_fn = "$sub_idx/test.$zp_idx.mol2.gz"

    return mol2_fn
end

@everywhere function AUC(x, y)
    """
    Calculates AUC using the trapezoidal rule,
    x = FPR
    y = TPR
    """

    s = length(x)

    a = zeros(s)
    for i in 1:s-1
        x1, x2 = x[i:i+1]
        y1, y2 = y[i:i+1]

        a[i] = (x2-x1) * ((y1 + y2) / 2)
    end

    return sum(a)
end

@everywhere function LogAUC(x, y, lambda = 0.001)
    """
    Calculated Adjusted LogAUC as decribed in Mysinger/Shoichet 2010
    doi://10.1021/ci100214a
    """

    logauc_max = 1.0
    random_logauc = (logauc_max-lambda)/log.(10)/log10.(logauc_max/lambda)

    # identifies starting point for arrays (first nonzero FPR)
    start = 0
    for i in 1:length(x)
        if x[i] >= lambda
            start = i
            break
        end
    end

    # slice arrays to correct size
    new_x = x[start:end]
    new_y = y[start:end]

    # take log of x
    log_x = log10.(new_x)

    # center log10 of x
    log_x = log_x .+ abs(min(log_x...))

    # bound log_x to 0,1
    log_x = log_x ./ max(log_x...)

    # calculate AUC and adjust for random_logauc
    log_auc = AUC(log_x, new_y) - random_logauc

    return log_auc
end

@everywhere function ReadName(name_fn)
    names = []
    open(name_fn) do io
        while !eof(io)
            line = readline(io)
            push!(names, line)
        end
    end
    return Set(names)
end

@everywhere function label!(frame, l_set, d_set)
    frame[!, :Type] = map(
        x -> if in(x, l_set) "ligand" elseif in(x, d_set) "decoy" else missing end,
        frame[!, :mol_name]
    )
end

@everywhere function ROC(frame)
    """
    Requires frame with :Type annotated as ligand/decoy

    FPR : False Positive Rate (FP / FP + TN)
    TPR : True Positive Rate (TP / TP + FN)
    """

    roc = Dict(:FPR => Float64[0.0], :TPR => Float64[0.0])

    FP = 0
    TP = 0

    P = sum(frame[:, :Type] .== "ligand")
    N = sum(frame[:, :Type] .== "decoy")

    for t in frame[:, :Type]
        if t == "ligand"
            TP += 1
        else
            FP += 1
        end

        fpr = FP / N
        tpr = TP / P

        push!(roc[:FPR], fpr)
        push!(roc[:TPR], tpr)

    end

    return DataFrame(roc, copycols=false)
end

@everywhere function CalculateAUC(score_frame, ligand_set, decoy_set)

    label!(score_frame, ligand_set, decoy_set)
    sort!(score_frame, :Total)

    roc = ROC(score_frame)
    auc = AUC(roc[!, :FPR], roc[!, :TPR])
    log_auc = LogAUC(roc[!, :FPR], roc[!, :TPR])

    return [@sprintf "%.4f" i*100 for i in [auc, log_auc]]
end

@everywhere function AddMolecule(line_name, io, mol_set, sub_idx, cls_idx, coord_frame)
    values = [sub_idx, cls_idx]

    name = replace(strip(line_name), re_MolName => "")

    # skip 4
    [readline(io) for i in 1:4]
    number = replace(strip(readline(io)), re_Number => "")

    # exit if number is not found
    if !in(number, mol_set)
        return false

    else
        # skip 4
        [readline(io) for i in 1:4]

        # prepare OXR
        oxr = []
        for i in 1:4
            line = split(replace(readline(io), re_OXR => ""))
            append!(oxr, line)
        end

        # skip 14
        [readline(io) for i in 1:14]

        # parse Total
        total = replace(strip(readline(io)), re_Total=>"")


        append!(values, [name, number, total])
        append!(values, oxr)

        push!(coord_frame, values)

        return true
    end
end

@everywhere function ParseMol2(sub_idx, cls_idx, mol_frame, coord_frame)

    gz_fn = build_gzfn(sub_idx, cls_idx)
    mol_set = Set(mol_frame.mol_idx)

    GZip.open(gz_fn) do io
        while !eof(io)
            line = readline(io)
            if occursin(re_MolName, line)
                AddMolecule(line, io, mol_set, sub_idx, cls_idx, coord_frame)
            end
        end
    end
end

@everywhere function ParseOUTDOCK(subcluster, score_frame, time_frame)

    # define regex expressions
    re_header = r"mol#"
    re_elapsed = r"elapsed"
    re_values = r"^ +[0-9]"
    re_ignore = r"^ 9|colors|skip_size|poses|<|>|no_match|bump|clashes"
    re_timesub = r"(elapsed.+\(sec\): +| +\(hour\).+)"

    # define file name
    OUTDOCK = "$subcluster/OUTDOCK"

    # start file parse
    current_group = 0
    open(OUTDOCK) do io
        while !eof(io)
            line = readline(io)

            # new group found
            if occursin(re_header, line)
                current_group += 1

            elseif occursin(re_values, line)
                if occursin(re_ignore, line)
                    continue
                end
                values = split(line)
                prepend!(values, [subcluster, string(current_group)])
                push!(score_frame, values)

            # final line found
            elseif occursin(re_elapsed, line)
                time = replace(line, re_timesub => "")
                push!(time_frame, [subcluster, time])
            end

        end
    end
end

@everywhere function ProcessDir(dir_name, q)
    println(dir_name)

    # initialize dataframes
    score_frame = DataFrame(
        sub_idx = String[],
        cls_idx = String[],
        mol_idx = String[],
        mol_name = String[],
        flex_code = String[],
        matched = String[],
        nscored = String[],
        time = String[],
        hac = String[],
        setnum = String[],
        matnum = String[],
        rank = String[],
        cloud = String[],
        elect = String[],
        gist = String[],
        vdW = String[],
        psol = String[],
        asol = String[],
        inter = String[],
        rec_e = String[],
        rec_d = String[],
        r_hyd = String[],
        Total = String[]
    )
    time_frame = DataFrame(
        sub_idx = [],
        time = []
    )
    coord_frame = DataFrame(
        sub_idx = String[],
        cls_idx = String[],
        mol_name = String[],
        mol_idx = String[],
        mol_total = String[],
        OXR_X1 = String[],
        OXR_Y1 = String[],
        OXR_Z1 = String[],
        OXR_X2 = String[],
        OXR_Y2 = String[],
        OXR_Z2 = String[],
        OXR_X3 = String[],
        OXR_Y3 = String[],
        OXR_Z3 = String[],
        OXR_X4 = String[],
        OXR_Y4 = String[],
        OXR_Z4 = String[]
    )

    # generate sdi subcluster list
    subcluster_list = [
        joinpath(dir_name, i) for i in readdir(dir_name) if occursin("subcluster", i)
    ]

    # generate ligand/decoy filenames
    ligand_fn = joinpath(dir_name, "ligands.names")
    decoy_fn = joinpath(dir_name, "decoys.names")

    # generate ligand/decoy molecule name sets
    ligand_set = ReadName(ligand_fn)
    decoy_set = ReadName(decoy_fn)

    # parse each subcluster OUTDOCK
    for s in subcluster_list
        ParseOUTDOCK(s, score_frame, time_frame)
    end

    # convert types
    time_frame[!, :time] = map(x -> parse(Float64, x), time_frame[!, :time])
    score_frame[!, :Total] = map(x -> parse(Float64, x), score_frame[!, :Total])

    # sort on time and score/molecule
    sort!(time_frame, :time)
    sort!(score_frame, [:Total, :mol_name])

    # take best scoring unique molecules
    unique_score_frame = unique(score_frame, [:mol_name])

    # write dataframes
    CSV.write(
        "$dir_name/extract_all.sort.txt",
        select(score_frame, Not(:cls_idx)),
        delim="\t", header=false
        )
    CSV.write(
        "$dir_name/extract_all.sort.uniq.txt",
        select(unique_score_frame, Not(:cls_idx)),
        delim="\t", header=false
        )

    # Calculate AUC and LogAUC
    auc, log_auc = CalculateAUC(unique_score_frame, ligand_set, decoy_set)
    time_frame[!, :AUC] .= auc
    time_frame[!, :LogAUC] .= log_auc

    # calculate percentile
    pc = quantile!(unique_score_frame.Total, q)

    # filter scores for hit molecules
    top_score_frame = filter(x -> x.Total < pc, unique_score_frame)
    sort!(top_score_frame, [:sub_idx, :cls_idx, :mol_idx])

    # parse mol2 gzipped files for coordinates
    gdf = groupby(top_score_frame, [:sub_idx, :cls_idx])
    for (key, mol_frame) in pairs(gdf)
        ParseMol2(key.sub_idx, key.cls_idx, mol_frame, coord_frame)
    end

    # label type for ligand set
    label!(coord_frame, ligand_set, decoy_set)
    coord_frame[!, :mol_total] = map(
        x -> parse(Float64, x),
        coord_frame[!, :mol_total]
        )
    sort!(coord_frame, :mol_total)

    # write timing/enrichment dataframe
    CSV.write(
        "$dir_name/time_and_enrichment.tab",
        time_frame, delim="\t", header=false
        )

    # write coordinate dataframe
    CSV.write(
        "$dir_name/coords.tab",
        coord_frame, delim="\t"
    )
end

@everywhere function parallel_process(d, q)
    try
        ProcessDir(d, q)
    catch
        println("ERROR in {$d}")
        return
    end
end

function get_args()
    s = ArgParseSettings()

    @add_arg_table! s begin
        "--directories", "-i"
            help = "Directories to process"
            required = true
            nargs='+'
        "--quantile", "-q"
            help = "Quantile to filter scores (default = 0.1)"
            required = false
            default = 0.1
            arg_type = Float64
    end

    return parse_args(s)
end

function main()

    parsed_args = get_args()
    directories = parsed_args["directories"]

    arguments = [
        (d, parsed_args["quantile"]) for d in directories
    ]

    # Parallel Map Arguments
    pmap(
        (args)->parallel_process(args...),
        arguments
        )

end



main()

File: src/Extract.py
#!/usr/bin/env python3

import argparse
import glob
import os
import re
import sys
import gzip

import pandas as pd
from multiprocessing import Pool
from tqdm import tqdm

class Directory:
    def __init__(self, dirname):
        self.dirname = dirname
        self.subclusters = glob.glob(
            os.path.join(self.dirname, "subcluster*")
        )

        self.re_header = re.compile("^[ ]+mol#")
        self.re_values = re.compile("^[ ]+[0-9]")
        self.re_ignore = re.compile(
            "^ 9[0-9]+ 9[0-9]+|skip_size|colors|poses|<|>|no_match|bump|clashes"
            )

        self.re_first_ws = re.compile("^[ ]+")
        self.re_ws = re.compile("[ ]+")

        self.groups = []

        self.raw_mol2_data = []

    def process_line(self, line):
        line = self.re_first_ws.sub("", line)
        line = self.re_ws.sub("\t", line)
        tokens = line.strip().split("\t")

        tokens[0] = int(tokens[0])

        for i in range(2, len(tokens)):
            if '.' in tokens[i]:
                tokens[i] = float(tokens[i])
            else:
                tokens[i] = int(tokens[i])

        return tokens

    def process_outdock(self, fn, sc_id):
        f = open(fn, "r")

        cluster_id=0

        while True:
            try:
                line = next(f)

                if self.re_header.search(line):
                    cluster_id += 1

                elif self.re_ignore.search(line):
                    continue

                elif self.re_values.search(line):
                    tokens = [sc_id, cluster_id] + self.process_line(line)
                    self.groups.append(tokens)

            except StopIteration:
                break

    def process_subcluster(self, sc):
        sc_id = int(sc[-4:])
        outdock_fn = os.path.join(sc, "OUTDOCK")
        self.process_outdock(outdock_fn, sc_id)

    def mol2_parse_hit(self, line, f):
        lines = [line] + [next(f) for _ in range(28)]
        return lines

    def read_mol2(self, sub_idx, cls_idx):
        fn = os.path.join(
            self.dirname,
            "subcluster{:04}".format(sub_idx),
            "test.{:04}.mol2.gz".format(cls_idx)
        )

        f = gzip.open(fn, "rt")

        re_lookup = [
            re.compile("###[ ]+Name:[ ]+{}".format(i))
            for i in self.mol_lookup[(sub_idx, cls_idx)]
        ]

        found = 0

        while True:
            try:
                line = next(f)
                for r in re_lookup:
                    if r.search(line):
                        data = self.mol2_parse_hit(line, f)
                        self.raw_mol2_data.append([sub_idx, cls_idx] + data)
                        found += 1

            except StopIteration:
                break

    def parse_mol2(self):
        def prep_coords(ox_set):
            return [[i for i in ox.strip().split(" ") if i != ""][-3:] for ox in ox_set]

        frame = []
        for d in self.raw_mol2_data:

            sub_idx = d[0]
            cls_idx = d[1]
            name = d[2].split(" ")[-1].strip()
            oxr_set = prep_coords(d[12:16])
            oxs_set = prep_coords(d[16:20])
            total = round(float(d[-1].split(" ")[-1].strip()), 2)

            for i in range(4):
                row = [sub_idx, cls_idx, name, float(total)] + \
                    oxr_set[i] + oxs_set[i]
                frame.append(row)


        self.mol2_frame = pd.DataFrame(
            frame,
            columns = \
                ["sub_idx", "cls_idx", "mol_name", "Total"] + \
                ["rec_{}".format(x) for x in ["x", "y", "z"]] + \
                ["lig_{}".format(x) for x in ["x", "y", "z"]]
            )

    def process_mol2(self):
        to_process = self.uniq_scores.\
            loc[:,['sub_idx', 'cls_idx', 'mol_name']].\
            drop_duplicates().\
            values

        self.mol_lookup = {}

        for i in to_process:
            sub_idx, cls_idx, mol_name = i
            file = (sub_idx, cls_idx)
            if file not in self.mol_lookup:
                self.mol_lookup[file] = []

            self.mol_lookup[file].append(mol_name)

        for f in self.mol_lookup:
            sub_idx, cls_idx = f
            self.read_mol2(sub_idx, cls_idx)



        self.parse_mol2()

        # select for hits
        self.mol2_hits = self.uniq_scores.\
            merge(
                self.mol2_frame,
                left_on = ['sub_idx', 'cls_idx', 'Total', "mol_name"],
                right_on = ['sub_idx', 'cls_idx', 'Total', "mol_name"]
                ).\
            loc[:, self.mol2_frame.columns]

    def write(self):

        # for enrichment
        self.scores.drop('cls_idx', axis = 1).to_csv(
            os.path.join(self.dirname, "extract_all.sort.txt"),
            sep = "\t", index=False, header=None
        )

        # for enrichment
        self.uniq_scores.drop('cls_idx', axis = 1).to_csv(
            os.path.join(self.dirname, "extract_all.sort.uniq.txt"),
            sep = "\t", index=False, header=None
        )


        # for downstream processing
        self.uniq_scores.to_csv(
            os.path.join(self.dirname, "unique_scores.tab"),
            sep = "\t", index=False
        )

        # for mol2 coordinates
        self.mol2_hits.to_csv(
            os.path.join(self.dirname, "unique_coords.tab"),
            sep = "\t", index=False
        )


    def run(self):

        # collect OUTDOCK information
        [self.process_subcluster(s) for s in self.subclusters]

        # aggregate scores into dataframe
        self.scores = pd.DataFrame(
            self.groups,
            columns = [
                "sub_idx", "cls_idx", "mol_idx", "mol_name", "flex_code",
                "matched", "nscored", "time", "hac", "setnum", "matnum",
                "rank", "cloud", "elect", "gist", "vdW", "psol", "asol",
                "inter", "rec_e", "rec_d", "r_hyd", "Total"
                ]
            ).sort_values("Total")
        self.scores = self.scores[~self.scores.mol_name.isna()]

        # take best scoring of unique molecules
        self.uniq_scores = self.scores.\
            sort_values("Total").\
            drop_duplicates("mol_name")


        # collect mol2 data
        self.process_mol2()


        # write data
        self.write()

def run_dir(d):
    dir = Directory(d)
    dir.run()

def get_args():
    p = argparse.ArgumentParser()
    p.add_argument(
        "-i", '--input', required=True, nargs = "+",
        help = "input directory(s) to process"
        )
    p.add_argument(
        "-n", "--threads", required=False,
        default = 8, type=int,
        help = 'number of threads to process on'
    )
    args = p.parse_args()

    return args

def main():
    args = get_args()

    p = Pool(args.threads)
    p.map(run_dir, tqdm(args.input, desc='Processing Directories'))
    p.close()


if __name__ == '__main__':
    main()
